### ISSUE TITLE
MongoDB

### ISSUE THREAD

**Issue by bmaschio** [jolie/jolie#71]
@jolie/developers 

Hi guys

I was wandering if any of you would be interested into the implementation of  Jolie Service Library  for MongoDB. Would be nice to have someone to exchange idea on how to do this stuff
Regards
Balint

**Comment by fmontesi** [comment]
I personally don't use MongoDB (yet? :-) ), but I guess that a driver for Jolie would be used by other people?
Since Postgresql came up with support for basically the same features, I tend to choose it over Mongo for stability reasons.

From a quick look at their documentation for Java, it looks like there are a few basic operations to offer (insert, find, delete, etc.). Since Mongo supports structured documents, these could just take generic Jolie values and store them as they are, yielding a nice integration. It's unclear how we could specify their constraints though, e.g., or/and logical clauses in find and such.

**Comment by bmaschio** [comment]
@fmontesi 
I suppose this is the catch22 of any new technology do we try to follow any new trend or do we stick on our guns and select one or two db technology and work only on those.I think MongoDB could be one of those DB to be supported   
I had a quick look too and I found some use of constrained search 
http://www.mkyong.com/mongodb/java-mongodb-query-document/ 
I suppose the first step would be how to define the Basic Insert, find, delete Jolie operation

**Comment by fmontesi** [comment]
Sure, but I fully support any library: As long as somebody needs it, it's useful. :-)

We can just distribute it separately like we started doing with the Database service, the new versions are now given in a separate repo. This will help with versioning.

Yes, the first steps would be implementing a way to connect to a MongoDB instance, and then these basic operations.

**Comment by bmaschio** [comment]
@fmontesi 

Well the XMAS holidays  are coming up and I  will be less busy with my new job ( paperwork more paperwork ) so I can pick up this. I will start with the interface definition and opening a new repository called MongoDB

**Comment by bmaschio** [comment]
@fmontesi 

Ok I menaged   to install the db and connect via Java to MongoDB engine. Looking at the Data model MongoDB seems to have 1:1 correspondence to the jolie variable structure

```javascript
type ClientType:void{
   .name: string
   .surname: string
   .addess:void{
         .street: string
         .city:string
         .streetNumber:int
    }
}

db.getCollection("Client").insertOne( new Document ("ClientType", new Document("address", new Document().append("street", "Backer st")).append("city", "London").append("streetNumber", 10) ).append("name", "Sherlok") .append("Surname", "Holmes")) ;

```
The very cool part of all of this is that was not necessary to define the Client Table ( collection) so it could be jolie driver the structure of the DB

**Comment by bmaschio** [comment]
@fmontesi 

hi there and  happy new year
I have done a first implementation of  the MongoDB jolie service 
with the
the following interface
```javascript
interface MongoDBInterface {
  RequestResponse: 
  connect (undefined)(undefined),
  query (undefined)(undefined)
}
```
I haven't yet set the operations types to make easyer the early debugging

So far the use of the  driver library has been quite straightforward. the only problem is the way the library reports  possible errors
One would think that  in case of error the drive would rise an exception: well what do I know the driver uses a LoggerClass to output on the console.
 
```javascript
gen 04, 2016 12:25:11 PM com.mongodb.diagnostics.logging.JULLogger log
INFORMAZIONI: Cluster created with settings {hosts=[localhost:27017], mode=SINGL
E, requiredClusterType=UNKNOWN, serverSelectionTimeout='30000 ms', maxWaitQueueS
ize=500}
```
I thought a could implement my own class 

```javascript
class JULLogger implements Logger {
```
and override the logging setting and use mine to work with Exception based model
Is this possible do you think? Have you ever work in this way? 

the second problem is how to represent query selection criteria in a non SQL driver
il the link below some example of query

http://www.tutorialspoint.com/mongodb/mongodb_query_document.htm

at the moment  my query operation  allows the user to select only the table

and the result looks like this



```javascript
responseq>>>>Value
        .row[0]
                .address[0]
                        .zipcode[0] = 10075 : java.lang.String
                        .coord[0] = -73.9557413 : java.lang.Double
                        .coord[1] = 40.7720266 : java.lang.Double
                        .street[0] = 2 Avenue : java.lang.String
                        .building[0] = 1480 : java.lang.String
                .restaurant_id[0] = 41704620 : java.lang.String
                .name[0] = Vella : java.lang.String
                .cuisine[0] = Italian : java.lang.String
                .borough[0] = Manhattan : java.lang.String
                .grades[0]
                        .date[0] = Wed Oct 01 02:00:00 CEST 2014 : java.lang.Str
ing
                        .score[0] = 11.0 : java.lang.Double
                        .grade[0] = A : java.lang.String
                .grades[1]
                        .date[0] = Thu Jan 16 01:00:00 CET 2014 : java.lang.Stri
ng
                        .score[0] = 17.0 : java.lang.Double
                        .grade[0] = B : java.lang.String
```

**Comment by bmaschio** [comment]
@fmontesi Are you ok?

**Comment by fmontesi** [comment]
@bmaschio Yes now. :-) Baby started in kindergarten, which decreased general health in the house for some weeks... :-)

Your work on mongo seems very promising! I've never tried overriding a logger but I guess if there's no other option that's the way to go. I find it very weird though that there is no other way to catch exceptions... go figure.

@jolie/developers any comments on this point and the other raised by @bmaschio ?

**Comment by bmaschio** [comment]
@fmontesi @jolie/developers 

I had a silly idea what about using a SQL Parser to allow our user to express Mongo DB  operation in familiar  manner 
A parser like 

http://zql.sourceforge.net/

What do you think

Balint

**Comment by fmontesi** [comment]
Hi Balint,

What do you mean? Can you show a simple example (doesn't need to be "real",
just a couple of lines to understand where you want to put SQL code) ?

- Fabrizio

**Comment by bmaschio** [comment]
@fmontesi 

Hi Fabrizio

Less start from  how MongoDB does things 

db.products.find() => select * from products

Now less look at a query with a condition 

db.products.find( { qty: { $gt: 25 } } ) => select * from products where qty > 25

Now I was wandering how to represent the 
```javascript
db.products.find( { qty: { $gt: 25 } } )
query@MongoDb(requestQueryType)(responseQueryType);

type requestQueryType:void{
.collectionName:string
.condition:void{
    .conditionVariable:string
    .conditionId:string
    .conditionValue: undefined
}
}

now let us try to express the condition

requestQuery.collectionName = "products";
requestQuery.condion.conditionVariable= "qty";
requestQuery.condion.conditionId = "$gt"
requestQuery.condion.conditionValue = 25 


```
I don't know you but i do not like this, so i would suggest to use a canonical SQL form 

```javascript

Now the query type will look lke this

 requestQueryType : string { ? }

requestQuery = select * from products where qty >:qty
request.qty= 25;
 
Now the Sql Parse will identify the diffrent component and transform it the MongoDb find 

db.products.find( { qty: { $gt: 25 } } )

```

What do you think?

Balint

**Comment by bmaschio** [comment]
@fmontesi 

Ok I did a bit of thinking, MongoDB is part of the not SQL DB ( see link  ) 

http://accionlabs.com/choosing-a-nosql-database-technology-comparison-matrix/

so perhaps using sql to express not sql expression it is a bit self defeating.  so   mayve ut woukkd be better to create a not sql parser

Balint

**Comment by bmaschio** [comment]
@fmontesi 

I think i found the way forward how to express things

        DBObject dbObject = (DBObject) JSON.parse("{ qty: { $gt: \"$qty\" } }");

using a JSON parser  it seem to be a good way

so now 

![image](https://cloud.githubusercontent.com/assets/12083716/13257490/44c2295a-da4f-11e5-9d8f-63c10259a3fe.png)

By using \"$qty\" we could inject the variable like we do in the normal sql

What do you think

**Comment by bmaschio** [comment]
@fmontesi 

Any comments?

**Comment by fmontesi** [comment]
Mmh, very interesting. Would it be possible maybe to just use Jolie values to express everything we need then? Just like they are doing with JSON.

Example:

find@Mongo( "products" {
    .qty.("$gt") = 25
} )( result )

If integrating with Javascript is easier, we could write the service in Javascript. There's a new shiny embedding system in the java8 branch that does automatic Jolie <-> JSON conversions.

**Comment by bmaschio** [comment]
@fmontesi 

Well it seems that MongoDB has select json as "language" to express query instead of SQL. I suppose one could create a Jolie representation of a not SQL (query ) but then we will need to create the mapping between Jolie representation and the  DBObject dbObject.
Furthermore I think would be counterproductive to move away from the MongoDB standard  representation it would be confusing for those developers that would like to use jolie to access MongoDB in micro service point of view.

**Comment by bmaschio** [comment]
@fmontesi @klag @jolie/developers 

Ok we got somewhere finally
Now  we got something kind of working  operation query look now something like

```javascript
No query
q.table = "prove";
```

![image](https://cloud.githubusercontent.com/assets/12083716/14322619/d306aa16-fc1d-11e5-9161-ce17a9b06fbc.png)



```javascript
Hard coded query

q.table = "prove";
q.query = "{nome: \"Tom\"}";
```

![image](https://cloud.githubusercontent.com/assets/12083716/14322517/7a09a972-fc1d-11e5-9cb3-8845018a503a.png)

```javascript
Passing the value via a variable
q.table = "prove";
q.query = "{nome: \"$nome\"}";
q.query.$nome = "Tom";
```
![image](https://cloud.githubusercontent.com/assets/12083716/14322653/eca9a2d4-fc1d-11e5-96c9-02719af5d7b3.png)

```javascript
Using logical operator
q.table = "prove";
q.query = "{age: { $gt: \"$age\"}}";
q.query.age = 32;
```

![image](https://cloud.githubusercontent.com/assets/12083716/14322927/db2f331a-fc1e-11e5-9a23-f6edcafebad2.png)

```javascript
Using array search
q.table = "prove";
q.query = "{nome: { $in:[\"$name1\",\"$name2\", \"$name3\"]}}";
q.query.name1 = "Tom";
q.query.name2 = "John";
q.query.name3 = "Balint";

I would prefare  use an array notation but I am working  on it
 
```
![image](https://cloud.githubusercontent.com/assets/12083716/14323206/e51f9652-fc1f-11e5-9e6e-bc1ba32ab57a.png)



What do you think 

Balint

**Comment by fmontesi** [comment]
This is really cool! :+1: 

@jolie/developers Is anybody else familiar with the MongoDB API? Do you find Balint's syntax familiar?

**Comment by bmaschio** [comment]
@fmontesi @jolie/developers 

My analysis 

https://docs.mongodb.org/getting-started/java/query/ 

then i decided to use a JSON representation that looked friendly enough

http://stackoverflow.com/questions/23362399/meteor-passing-mongodb-selector-from-parameter

Tnk

Balint

**Comment by bmaschio** [comment]
@fmontesi    @jolie/developers 

using the same login i have implemented the delete and update operation

**Comment by fmontesi** [comment]
Good thinking.

I'm just unsure of the terminology. Why "table" and not "collection"? How do people call them in MongoDB? We should try to be faithful to their terms, so that a person who knows MongoDB finds our API familiar.

I like that you reused the query parameter concepts from Database. :-)
But why do you use $param instead of what we use in Database, :param ?

**Comment by bmaschio** [comment]
@fmontesi 

Good catch yes you are right on talking about collection and not table being MongoDB a document base DB. 
I use $Parameter because the JSON Parser will not work otherwise like in the example. i was being leasy and use the library JSON parser instead of writing my own
I was thinking last week-end also how handy will come the  jolie "for each" in handling MongoDB data

**Comment by bmaschio** [comment]
@fmontesi 

On the renaming i would go like this

table => collection
data=> document
query=> filter

what you think?

**Comment by LarsIlver** [comment]
Hi.
I'm very interested in having MongoDB as part of the Jolie service library. New on this thread, my first input is regarding naming: I thing this https://docs.mongodb.org/manual/reference/sql-comparison/ will be a good foundation for interface definitions.

**Comment by bmaschio** [comment]
@LarsIlver @fmontesi 

HI Larsllver, first of all thank for your help,  I have started to implement a bit of the Jolie MongoDB, 

```javascript
type queryRequest : void {
  .collection : string
  .query : string{
     undefined
  }
}

type queryResponse: void{
   .row*: string {
        undefined
   } //this is the mongo DB ID
 
}

q.collection= "test";
q.query = "{nome: \"$nome\"}";
q.query.nome = "Tom";
```
I used the JSON format as suggested by some example of query, what i do not like is the use of the tocken $.

For inserting it looks 


```javascript
type  insertRequest : void {
  .collection : string
  .document *: undefined
}

type queryResponse: void{
   .id*: string
}

```
For deleting 


```javascript
type  deleteRequest : void {
  .collection: sting 
  .id? : string
  .query? : string{
     undefined
   }
 
}

```

for updating

```javascript
type  updateRequest : void {
  .collection: sting 
  .id? : string
  .query? : string{
     undefined
   }
  .document*:undefined
 
}

```

How do you feel about it ? 

Thank you
Balint

**Comment by LarsIlver** [comment]
Hi Balint.

Looks fine, but I think you miss the top levels.
-host
-port (could default to 27019, but can be anything)
-db

When host, port and collection has been defined you can go on to CRUD at some collection

I am a little busy these days, hope you can accept that I will return with some more constructive in a few days.

/Lars




Fra: Balint Maschio [mailto:notifications@github.com]
Sendt: 2. maj 2016 09:56
Til: jolie/jolie <jolie@noreply.github.com>
Cc: Lars Johan Ilver <LJI@odense.dk>; Mention <mention@noreply.github.com>
Emne: Re: [jolie/jolie] MongoDB (#71)


@LarsIlver<https://github.com/LarsIlver> @fmontesi<https://github.com/fmontesi>

HI Larsllver, first of all thank for your help, I have started to implement a bit of the Jolie MongoDB,

type queryRequest : void {

  .collection : string

  .query : string{

     undefined

  }

}



type queryResponse: void{

   .row*: string {

        undefined

   } //this is the mongo DB ID



}



q.collection= "test";

q.query = "{nome: \"$nome\"}";

q.query.nome = "Tom";

I used the JSON format as suggested by some example of query, what i do not like is the use of the tocken $.

For inserting it looks

type  insertRequest : void {

  .collection : string

  .document *: undefined

}



type queryResponse: void{

   .id*: string

}

For deleting

type  deleteRequest : void {

  .collection: sting

  .id? : string

  .query? : string{

     undefined

   }



}

for updating

type  updateRequest : void {

  .collection: sting

  .id? : string

  .query? : string{

     undefined

   }

  .document*:undefined



}

How do you feel about it ?

Thank you
Balint

**Comment by LarsIlver** [comment]
Hi Balint,
A correction to line 6 (db)

Fra: Lars Johan Ilver
Sendt: 3. maj 2016 19:52
Til: 'jolie/jolie' <reply@reply.github.com>
Emne: SV: [jolie/jolie] MongoDB (#71)

Hi Balint.

Looks fine, but I think you miss the top levels.
-host
-port (could default to 27019, but can be anything)
-db

When host, port and db has been defined you can go on to CRUD at some collection

I am a little busy these days, hope you can accept that I will return with some more constructive in a few days.

/Lars




Fra: Balint Maschio [mailto:notifications@github.com]
Sendt: 2. maj 2016 09:56
Til: jolie/jolie <jolie@noreply.github.com<mailto:jolie@noreply.github.com>>
Cc: Lars Johan Ilver <LJI@odense.dk<mailto:LJI@odense.dk>>; Mention <mention@noreply.github.com<mailto:mention@noreply.github.com>>
Emne: Re: [jolie/jolie] MongoDB (#71)


@LarsIlver<https://github.com/LarsIlver> @fmontesi<https://github.com/fmontesi>

HI Larsllver, first of all thank for your help, I have started to implement a bit of the Jolie MongoDB,

type queryRequest : void {

  .collection : string

  .query : string{

     undefined

  }

}



type queryResponse: void{

   .row*: string {

        undefined

   } //this is the mongo DB ID



}



q.collection= "test";

q.query = "{nome: \"$nome\"}";

q.query.nome = "Tom";

I used the JSON format as suggested by some example of query, what i do not like is the use of the tocken $.

For inserting it looks

type  insertRequest : void {

  .collection : string

  .document *: undefined

}



type queryResponse: void{

   .id*: string

}

For deleting

type  deleteRequest : void {

  .collection: sting

  .id? : string

  .query? : string{

     undefined

   }



}

for updating

type  updateRequest : void {

  .collection: sting

  .id? : string

  .query? : string{

     undefined

   }

  .document*:undefined



}

How do you feel about it ?

Thank you
Balint

**Comment by bmaschio** [comment]
@LarsIlver @fmontesi 
Hi Lars

Yes no problem I am looking forward to talk to you about MongoDB
In the meantime let me keep you posted on some of the development
You correctly sported the absence of  the connection information, that is because I have decided (  an arbitrary decision) to define the connection parameters in connect operation. This will be used at micro-service level. So that the connection would be establish only once at the start of the microservice
It also make sense in my head ( and that really means little ) to have a highly specialized microservice for each DB
I am interested in investigating the possibility to use Jolie and the microservices architecture to implement  a data federator. In particular under a BI framework, I like very much the document based approach of MongoDB, I think is excellent when approcing data from a Business point of view.
On the more implementation side of things, I have used the java libraries to implement the jolie native service. Yesterday I had the unhealthy idea to go through  with a driver update from 3.0.2 to 3.2.2, and with my surprise some of the implementation had to be recoded.
Yet I still have some perplexities on the implementation

1. How to inject variables into NO SQL Filters 

2.How to deal with the update  

3. How to represent the id

Regards
Balint

**Comment by LarsIlver** [comment]
Hi Balint.

Agree that connect operation should be handled on micro-service level.
(I noticed that you categorized test as a collection, it’s the default db when connect, that’s why I listed all levels)
I think highly specialized m-services is a good approach to support m-service architechture in general.

My need for NoSQL could be compared to data federation principles. That is, the flexibility for mixing different datastructures in a context that is IOT ‘ish (Smart City solutions).

I take a closer look at the concerns you list and will revert.

/Lars


Fra: Balint Maschio [mailto:notifications@github.com]
Sendt: 4. maj 2016 09:32
Til: jolie/jolie <jolie@noreply.github.com>
Cc: Lars Johan Ilver <LJI@odense.dk>; Mention <mention@noreply.github.com>
Emne: Re: [jolie/jolie] MongoDB (#71)


@LarsIlver<https://github.com/LarsIlver> @fmontesi<https://github.com/fmontesi>
Hi Lars

Yes no problem I am looking forward to talk to you about MongoDB
In the meantime let me keep you posted on some of the development
You correctly sported the absence of the connection information, that is because I have decided ( an arbitrary decision) to define the connection parameters in connect operation. This will be used at micro-service level. So that the connection would be establish only once at the start of the microservice
It also make sense in my head ( and that really means little ) to have a highly specialized microservice for each DB
I am interested in investigating the possibility to use Jolie and the microservices architecture to implement a data federator. In particular under a BI framework, I like very much the document based approach of MongoDB, I think is excellent when approcing data from a Business point of view.
On the more implementation side of things, I have used the java libraries to implement the jolie native service. Yesterday I had the unhealthy idea to go through with a driver update from 3.0.2 to 3.2.2, and with my surprise some of the implementation had to be recoded.
Yet I still have some perplexities on the implementation

  1.  How to inject variables into NO SQL Filters

2.How to deal with the update

  1.  How to represent the id

Regards
Balint

**Comment by bmaschio** [comment]
@LarsIlver @fmontesi 

I got somewhere with my coding
Here is the interface 
```javascript
type ConnectRequest:void{
  .host : string
  .dbname : string
  .port :int
}

type ConnectResponse: void

type QueryRequest:void{
   .collection: string
   .filter?:undefined
}

type QueryResponse:void{
   .document*: undefined
}
type InsertRequest:void{
  .collection: string
  .document:undefined
}

type InsertResponse:void


type UpdateRequest:void{
  .collection: string
  .filter:undefined
  .documentUpdate:undefined
}

type UpdateResponse:void

type DeleteRequest:void{
  .collection: string
  .filter?:undefined
}

type DeleteResponse:void

interface MongoDBInterface {
  RequestResponse:
  connect (ConnectRequest)(ConnectResponse) throws MongoException ,
  query   (QueryRequest)(QueryResponse)   throws MongoException JsonParseException ,
  insert  (InsertRequest)(InsertResponse)   throws MongoException JsonParseException ,
  update  (UpdateRequest)(UpdateResponse)   throws MongoException JsonParseException ,
  delete  (DeleteRequest)(DeleteResponse)   throws MongoException JsonParseException
}


outputPort MongoDB {
Interfaces: MongoDBInterface
}

embedded {
Java:
	"joliex.mongodb.MongoDbConnector" in MongoDB
}

```

The connect operation needs some work on adding credential handling 
the use the insert operation
```javascript
q.collection = "CustomerSales";
with (q.document){
      .name    = "Lars";
      .surname = "Larsesen";
      .code = "LALA01";
      .age = 28;
      with (.purchase){
        .ammount = 30.12;
        .date = "12.03.2016"
      }
    };
valueToPrettyString@StringUtils (q)(s);
println@Console("q>>>>"+s)();
insert@MongoDB(q)(responseq);
valueToPrettyString@StringUtils (responseq)(s);
println@Console("responseq>>>>"+s)();
undef(q)
```

running it looks like

![image](https://cloud.githubusercontent.com/assets/12083716/15221880/8cee1b06-186d-11e6-8683-c83f0e0c2751.png)

the query operation looks like 

```javascript
q.collection = "CustomerSales";
/*q.query = "{\"spesa.ammount\":{$gt:\"$ammount\"}}";*/
q.filter = "{\"age\":{$gt:\"$age\"}}";
q.filter.age = 20;
```

the results looks like

![image](https://cloud.githubusercontent.com/assets/12083716/15221987/050690aa-186e-11e6-8804-3772bea00c94.png)

the upgrade code looks like

```javascript
q.collection = "CustomerSales";
q.filter = "{name: \"$name1\"}";
q.filter.name1 = "Balint";
q.documentUpdate = "{$set:{age:\"$age\"}}";
q.documentUpdate.age= 38;
```

with a result that looks like 

![image](https://cloud.githubusercontent.com/assets/12083716/15222359/8e1ceb9a-186f-11e6-84d3-7316c4673513.png)

I kind of had an idea to handle the datetime not native in jolie
```javascript

q.document,dateRecoding = "01.01.2016"
q.document,dateRecoding.@DateTime.format ="DD:MM:YYYY"
q.document,dateRecoding.@DateTime.lang ="it"
```
Soon I will try to implement an aggregation operation

What do you think

Balint

**Comment by LarsIlver** [comment]
Hi Balint.

I think it looks great!!

Looking forward to try it out.

Regards Lars

Fra: Balint Maschio [mailto:notifications@github.com]
Sendt: 12. maj 2016 18:35
Til: jolie/jolie <jolie@noreply.github.com>
Cc: Lars Johan Ilver <LJI@odense.dk>; Mention <mention@noreply.github.com>
Emne: Re: [jolie/jolie] MongoDB (#71)


@LarsIlver<https://github.com/LarsIlver> @fmontesi<https://github.com/fmontesi>

I got somewhere with my coding
Here is the interface

type ConnectRequest:void{

  .host : string

  .dbname : string

  .port :int

}



type ConnectResponse: void



type QueryRequest:void{

   .collection: string

   .filter?:undefined

}



type QueryResponse:void{

   .document*: undefined

}

type InsertRequest:void{

  .collection: string

  .document:undefined

}



type InsertResponse:void





type UpdateRequest:void{

  .collection: string

  .filter:undefined

  .documentUpdate:undefined

}



type UpdateResponse:void



type DeleteRequest:void{

  .collection: string

  .filter?:undefined

}



type DeleteResponse:void



interface MongoDBInterface {

  RequestResponse:

  connect (ConnectRequest)(ConnectResponse) throws MongoException ,

  query   (QueryRequest)(QueryResponse)   throws MongoException JsonParseException ,

  insert  (InsertRequest)(InsertResponse)   throws MongoException JsonParseException ,

  update  (UpdateRequest)(UpdateResponse)   throws MongoException JsonParseException ,

  delete  (DeleteRequest)(DeleteResponse)   throws MongoException JsonParseException

}





outputPort MongoDB {

Interfaces: MongoDBInterface

}



embedded {

Java:

    "joliex.mongodb.MongoDbConnector" in MongoDB

}

The connect operation needs some work on adding credential handling
the use the insert operation

q.collection = "CustomerSales";

with (q.document){

      .name    = "Lars";

      .surname = "Larsesen";

      .code = "LALA01";

      .age = 28;

      with (.purchase){

        .ammount = 30.12;

        .date = "12.03.2016"

      }

    };

valueToPrettyString@StringUtils (q)(s);

println@Console("q>>>>"+s)();

insert@MongoDB(q)(responseq);

valueToPrettyString@StringUtils (responseq)(s);

println@Console("responseq>>>>"+s)();

undef(q)

running it looks like

[image]<https://cloud.githubusercontent.com/assets/12083716/15221880/8cee1b06-186d-11e6-8683-c83f0e0c2751.png>

the query operation looks like

q.collection = "CustomerSales";

/*q.query = "{\"spesa.ammount\":{$gt:\"$ammount\"}}";*/

q.filter = "{\"age\":{$gt:\"$age\"}}";

q.filter.age = 20;

the results looks like

[image]<https://cloud.githubusercontent.com/assets/12083716/15221987/050690aa-186e-11e6-8804-3772bea00c94.png>

the upgrade code looks like

q.collection = "CustomerSales";

q.filter = "{name: \"$name1\"}";

q.filter.name1 = "Balint";

q.documentUpdate = "{$set:{age:\"$age\"}}";

q.documentUpdate.age= 38;

with a result that looks like

[image]<https://cloud.githubusercontent.com/assets/12083716/15222359/8e1ceb9a-186f-11e6-84d3-7316c4673513.png>

I kind of had an idea to handle the datetime not native in jolie



q.document,dateRecoding = "01.01.2016"

q.document,dateRecoding.@DateTime.format ="DD:MM:YYYY"

q.document,dateRecoding.@DateTime.lang ="it"

Soon I will try to implement an aggregation operation

What do you think

Balint

**Comment by bmaschio** [comment]
@fmontesi @LarsIlver 

More updates Mongo World

I have added a debug flag to the connection that allows us to see how the bsonDocument is populated 

```javascript

type ConnectRequest:void{
  .host : string
  .dbname : string
  .port :int
  .jsonStringDebug?:bool
}

```
result

![image](https://cloud.githubusercontent.com/assets/12083716/15252769/cb66d798-192f-11e6-9f10-08b84b3d7362.png)


I also discover that we do need to use " " to pass parameters so we can write stuff like 

```javascript
q.collection = "CustomerSales";
q.filter = "{name: '$name1'}";
q.filter.name1 = "Balint";
q.documentUpdate = "{$set:{age:'$age'}}";
q.documentUpdate.age= 70;
```
I have also implemented the aggregate operation

```javascript
type AggregateRequest:void{
    .collection: string
    .filter*:string{
       ?
    }
}

type AggregateResponse:void{
    .document*:undefined
}
interface MongoDBInterface {
  RequestResponse:
  connect (ConnectRequest)(ConnectResponse) throws MongoException ,
  query   (QueryRequest)(QueryResponse)   throws MongoException JsonParseException ,
  insert  (InsertRequest)(InsertResponse)   throws MongoException JsonParseException ,
  update  (UpdateRequest)(UpdateResponse)   throws MongoException JsonParseException ,
  delete  (DeleteRequest)(DeleteResponse)   throws MongoException JsonParseException ,
  aggregate (AggregateRequest)(AggregateResponse)   throws MongoException JsonParseException
}
```
so we can write stuff like 

```javascript

q.collection = "CustomerSales";
q.filter = "{$group:{ _id : '$name', total:{$sum : 1}}}";

```

with result 

![image](https://cloud.githubusercontent.com/assets/12083716/15253684/aece111a-1933-11e6-9fca-eb044155e31c.png)


@fmontesi can we discuss on timestamp issue?

Regards 

Balint

**Comment by fmontesi** [comment]
Hi Balint,

Great work!

Can you explain what the problem is with the timestamps/dates, exactly?

- F

**Comment by bmaschio** [comment]
@fmontesi 

Not having a native Data native type in jolie i thought to use the following complex type 

```javascript

type DateTime: string{
         .@DateTimeFormat: string
         .@DateTimeLang?: string
}

```
I would use it in this way

```javascript

request.ExpDate = "22/10/2016"
request.ExpDate.@DateTimeFormat="DD/MM/YYYY"
request.ExpDate..@DateTimeLang = "it"
```


We could also use the same idea for coordinate type

```javascript
type CoordinateType: string {
       .@CordSeparator: String
}

```

I would use it in this way

```javascript

request.cood = "33.934245, -0.527344"
request.cood@CordSeparator=","
```
What do you think ?

**Comment by fmontesi** [comment]
It seems that Mongo uses standard JavaScript Date objects. Those are documented here:

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date

Some fields are optional. This roughly translates to:

```
type Date:void {
    .year:int
    .month:int
    .day?:int
    .hour?:int
    .minute?:int
    .second?:int
    .millisecond?:int
}
```

So that's what we should pass to the operations. I would avoid having automatic transformation algorithms in the core operations (update, etc.), to avoid complexity. Instead, I would put some convenience transformer operations in the interface, like stringToDate, dateToString, etc.

**Comment by bmaschio** [comment]
@fmontesi 

the BSON object accept a java long type so you would transform that structure in a long?

```
BsonDateTime objToInsert= new BsonDateTime(port)
```
 is that correct?

**Comment by fmontesi** [comment]
Yes.

By the way, looking through your code, it seems that you disregard sub-sub-...-sub-nodes in trees.

For example, would the following document be OK for insertion?

```
{
   .person.firstName = "Homer";
   .person.lastName = "Simpson"
}
```

**Comment by bmaschio** [comment]
if (valueVector.get(counterValueVector).hasChildren()) {

                        bsonArray.add(counterValueVector, createDocument(valueVector.get(counterValueVector)));

                    }

this should be recursive?

Am I wronge?

**Comment by fmontesi** [comment]
Oh, my bad, I missed that.

- F

**Comment by bmaschio** [comment]
```
with (q.document){
      .name    = "Lars";
      .surname = "Larsesen";
      .code = "LALA01";
      .age = 28;
      with (.purchase){
        .ammount = 30.12;
        .date.("@type")="Date";
        .date= long(170000);
        .location.street= "Mongo road";
          .location.number= 2
      }
    };
```
 result of the query
![image](https://cloud.githubusercontent.com/assets/12083716/15319021/10bfa08a-1c29-11e6-9cb8-fd895083699b.png)

How do you like it?

**Comment by bmaschio** [comment]
@fmontesi 

any thought on this?

**Comment by bmaschio** [comment]
@fmontesi 

I just found a bit of a problem with DateTime type it does save it always in UTC I found a solution  but i do not like particularly

http://stackoverflow.com/questions/31654567/mongodb-bson-stores-date-in-utc-time

It entails the use of this extra library

https://sourceforge.net/projects/joda-time/?source=typ_redirect

What do you think?

**Comment by fmontesi** [comment]
Joda time is well known, go ahead. I don't mind having a dependency here since we'll probably distribute this module separately. :-)

**Comment by fmontesi** [comment]
Hi @bmaschio, can we close this? I guess it's fixed by the mongo db connector distributed separately?
